bot/exts/info/doc/_parsing.py

```diff
@@ -0,0 +1,328 @@
+from __future__ import annotations
+
+import logging
+import re
+import string
+import textwrap
+from functools import partial
+from typing import Callable, Collection, Container, Iterable, List, Optional, TYPE_CHECKING, Union
+
+from bs4 import BeautifulSoup
+from bs4.element import NavigableString, PageElement, Tag
+
+from bot.utils.helpers import find_nth_occurrence
+from ._html import Strainer
+from ._markdown import DocMarkdownConverter
+if TYPE_CHECKING:
+    from ._cog import DocItem
+
+log = logging.getLogger(__name__)
+
+_MAX_SIGNATURE_AMOUNT = 3
+
+_UNWANTED_SIGNATURE_SYMBOLS_RE = re.compile(r"\[source]|\\\\|¶")
+_WHITESPACE_AFTER_NEWLINES_RE = re.compile(r"(?<=\n\n)(\s+)")
+_PARAMETERS_RE = re.compile(r"\((.+)\)")
+
+_SEARCH_END_TAG_ATTRS = (
+    "data",
+    "function",
+    "class",
+    "exception",
+    "seealso",
+    "section",
+    "rubric",
+    "sphinxsidebar",
+)
+
+_NO_SIGNATURE_GROUPS = {
+    "attribute",
+    "envvar",
+    "setting",
+    "tempaltefilter",
+    "templatetag",
+    "term",
+}
+_EMBED_CODE_BLOCK_LINE_LENGTH = 61
+# _MAX_SIGNATURE_AMOUNT code block wrapped lines with py syntax highlight
+_MAX_SIGNATURES_LENGTH = (_EMBED_CODE_BLOCK_LINE_LENGTH + 8) * _MAX_SIGNATURE_AMOUNT
+# Maximum discord message length - signatures on top
+_MAX_DESCRIPTION_LENGTH = 2000 - _MAX_SIGNATURES_LENGTH
+_TRUNCATE_STRIP_CHARACTERS = "!?:;." + string.whitespace
+_BRACKET_PAIRS = {
+    "{": "}",
+    "(": ")",
+    "[": "]",
+}
+
+
+def _split_parameters(parameters_string: str) -> List[str]:
+    """
+    Split parameters of a signature into individual parameter strings on commas.
+
+    Long string literals are not accounted for.
+    """
+    parameters_list = []
+    last_split = 0
+    depth = 0
+    expected_end = None
+    current_search = None
+
+    for index, character in enumerate(parameters_string):
+        if character in _BRACKET_PAIRS:
+            if current_search is None:
+                current_search = character
+                expected_end = _BRACKET_PAIRS[character]
+            if character == current_search:
+                depth += 1
+
+        elif character in {"'", '"'}:
+            if depth == 0:
+                depth += 1
```

Comment:
That shouldn't really be a problem because we're ignoring commas to split on anyway if we're within brackets, however the parsing fails when we decrease depth on string matches below without knowing if the initial depth increase came from a string or one of the brackets